### PR TITLE
Automatic update for HTTPS proxy TLS certificates

### PR DESCRIPTION
**Pull request by koct9i**
- Add methods for SSL context update
  This is required for TLS certificate rotation without downtime.
  
- Automatically update HTTPS server TLS certificates
  
- yt/environmnt: keep TLS certificates in files
  
- Add test for http proxy certificate update

### CODE REVIEW

File: yt/yt/tests/integration/proxies/test_http_proxy.py [hunk 1]
```diff
@@ -1154,3 +1177,52 @@ def test_ping_https_verify_false(self):
...
+    def test_certificate_broken_update(self):
+        proxy_cert, proxy_cert_key = self._get_proxy_cert_path()
+
+        with open(proxy_cert, 'rb') as f:
+            cert_content = f.read()
+
+        with open(proxy_cert, 'w') as f:
+            f.write('hello')
+
+        time.sleep(3)
```
Comment (Kontakter) [inline yt/yt/tests/integration/proxies/test_http_proxy.py:1221]: It is better to use `wait` function

File: yt/yt/tests/integration/proxies/test_http_proxy.py [hunk 2]
```diff
@@ -1154,3 +1177,52 @@ def test_ping_https_verify_false(self):
...
+
+    @authors("khlebnikov")
+    def test_certificate_broken_update(self):
+        proxy_cert, proxy_cert_key = self._get_proxy_cert_path()
+
+        with open(proxy_cert, 'rb') as f:
+            cert_content = f.read()
+
+        with open(proxy_cert, 'w') as f:
+            f.write('hello')
```
Comment (Kontakter) [inline yt/yt/tests/integration/proxies/test_http_proxy.py:1219]: Please use double quotes

File: yt/yt/tests/integration/proxies/test_http_proxy.py (hunk 1 above)
Comment (koct9i) [reply yt/yt/tests/integration/proxies/test_http_proxy.py]: I dont'n see how to do that without adding structured logging events for that case. This 
seems too overengineed.

I've removed this test case. It's not so important anyway.

**Comment by yalekseev** [comment]
Shipped!

File: yt/yt/core/https/server.cpp [hunk 3]
```diff
@@ -22,8 +30,9 @@ class TServer
     : public IServer
 {
 public:
-    explicit TServer(IServerPtr underlying)
+    explicit TServer(IServerPtr underlying, TPeriodicExecutorPtr certificateUpdater)
```
Comment (maxim-babenko) [inline yt/yt/core/https/server.cpp:33]: explicit is now redundant

File: yt/yt/core/crypto/tls.h [hunk 4]
```diff
@@ -48,6 +52,7 @@ class TSslContext
 
 private:
     const TIntrusivePtr<TSslContextImpl> Impl_;
+    TInstant CommitTime_;
```
Comment (maxim-babenko) [inline yt/yt/core/crypto/tls.h:55]: Why placing it here instead of hiding inside pimpl?

File: yt/yt/core/https/server.cpp [hunk 5]
```diff
@@ -62,28 +77,66 @@ class TServer
...
+    TPeriodicExecutorPtr certificateUpdater;
+    if (sslConfig->UpdatePeriod &&
+        sslConfig->CertChain->FileName &&
+        sslConfig->PrivateKey->FileName)
+    {
+        certificateUpdater = New<TPeriodicExecutor>(
+            poller->GetInvoker(),
+            BIND([=, serverName = config->ServerName] {
+                try {
+                    auto modificationTime = Max(
```
Comment (maxim-babenko) [inline yt/yt/core/https/server.cpp:120]: Now you have IO in poller thread, seems like a problem

File: yt/yt/core/https/server.cpp [hunk 6]
```diff
@@ -62,28 +77,66 @@ class TServer
...
+                        NFS::GetPathStatistics(*sslConfig->CertChain->FileName).ModificationTime,
+                        NFS::GetPathStatistics(*sslConfig->PrivateKey->FileName).ModificationTime);
+
+                    // Detect fresh and stable updates.
+                    if (modificationTime > sslContext->GetCommitTime() &&
+                        modificationTime + sslConfig->UpdatePeriod <= TInstant::Now())
+                    {
+                        YT_LOG_INFO("Updating TLS certificates (ServerName: %v, ModificationTime: %v)", serverName, modificationTime);
```
Comment (maxim-babenko) [inline yt/yt/core/https/server.cpp:128]: Let's place the arguments on separate lines

File: yt/yt/core/https/server.cpp [hunk 7]
```diff
@@ -62,28 +77,66 @@ class TServer
...
+
+    auto sslConfig = config->Credentials;
+    TPeriodicExecutorPtr certificateUpdater;
+    if (sslConfig->UpdatePeriod &&
+        sslConfig->CertChain->FileName &&
+        sslConfig->PrivateKey->FileName)
+    {
+        certificateUpdater = New<TPeriodicExecutor>(
+            poller->GetInvoker(),
+            BIND([=, serverName = config->ServerName] {
```
Comment (maxim-babenko) [inline yt/yt/core/https/server.cpp:118]: Why not just =? Capturing `config` is no big deal.

**Pull request closed without merging**